packages/core/src/lib/filter/v2/index.ts

```diff
@@ -0,0 +1,359 @@
+ * - https://github.com/status-im/go-waku/issues/245
+ * - https://github.com/status-im/nwaku/issues/948
+ */
+class FilterV2 extends BaseProtocol implements IFilterV2 {
+  options: ProtocolCreateOptions;
+  private subscriptions: Map<PeerIdStr, unknown[]>;
+  public pubSubTopic: string;
+
+  constructor(public libp2p: Libp2p, options?: ProtocolCreateOptions) {
+    super(
+      FilterV2Codecs.SUBSCRIBE,
+      libp2p.peerStore,
+      libp2p.getConnections.bind(libp2p)
+    );
+
+    this.libp2p.handle(FilterV2Codecs.PUSH, this.onRequest.bind(this));
+
+    this.options = options ?? {};
+    this.subscriptions = new Map();
+
+    const { pubSubTopic = DefaultPubSubTopic } = this.options;
+    this.pubSubTopic = pubSubTopic;
+  }
+
+  /**
+   * @param decoders Decoder or array of Decoders to use to decode messages, it also specifies the content topics.
+   * @param callback A function that will be called on each message returned by the filter.
+   * @param opts The FilterSubscriptionOpts used to narrow which messages are returned, and which peer to connect to.
+   * @returns Unsubscribe function that can be used to end the subscription.
+   */
+  async subscribe<T extends IDecodedMessage>(
+    decoders: IDecoder<T> | IDecoder<T>[],
+    callback: Callback<T>,
+    opts?: ProtocolOptions
+  ): Promise<SubscriptionReturn<"v2">> {
+    const decodersArray = Array.isArray(decoders) ? decoders : [decoders];
+
+    const contentTopics = Array.from(groupByContentTopic(decodersArray).keys());
+
+    const request = FilterSubscribeRpc.createSubscribeRequest(
+      this.pubSubTopic,
+      contentTopics
+    );
+
+    const peer = await this.getPeer(opts?.peerId);
+    const stream = await this.newStream(peer);
+
+    try {
+      const res = await pipe(
+        [request.encode()],
+        lp.encode,
+        stream,
+        lp.decode,
+        async (source) => await all(source)
+      );
+
+      const { statusCode, requestId, statusDesc } =
+        FilterSubscribeResponse.decode(res[0].slice());
+
+      if (statusCode < 200 || statusCode >= 300) {
+        throw new Error(
+          `Filter subscribe request ${requestId} failed with status code ${statusCode}: ${statusDesc}`
+        );
+      }
+
+      log(
+        "Subscribed to peer ",
+        peer.id.toString(),
+        "for content topics",
+        contentTopics
+      );
+    } catch (e) {
+      throw new Error(
+        "Error subscribing to peer: " +
+          peer.id.toString() +
+          " for content topics: " +
+          contentTopics +
+          ": " +
+          e
+      );
+    }
+
+    const subscription: Subscription<T> = {
+      callback,
+      decoders: decodersArray,
+      pubSubTopic: this.pubSubTopic,
+    };
+    this.subscriptions.set(
+      peer.id.toString(),
+      [...(this.subscriptions.get(peer.id.toString()) ?? []), subscription] ?? [
+        subscription,
+      ]
+    );
+
+    return {
```

Comment:
I think it'd be neater if this would simply return a class (let's says `FilterSubscription`) that is constructed with the peer id and the FilterV2 class instead of filling each property of the object.

Actually, such class may be a better API:


```
const subscription = await FilterV2.createSubscription(pubSubTopic, callback); // automatically selects a peer id if not passed

// add content topics
let res = await subscriptiion.subscribe(["a", "b", "c"]);

// remove content topics
res = await subscriptiion.unsubscribe(["a"]

// Check subscription
res = awaiot subscriptiion.ping()

// Unsubscribe all

res = await subscriptiion.unsubscribeAll();

```

The API is not perfect as decoder needs to be used.

https://github.com/waku-org/js-waku/pull/1332/files#r1189231005

Reply:
Iterative change.